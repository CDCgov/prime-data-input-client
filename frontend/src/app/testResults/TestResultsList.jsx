import React from "react";
import moment from "moment";
import { useSelector } from "react-redux";
import { PATIENT_TERM_CAP } from "../../config/constants";

import { getTestResultsWithPatientDetails } from "./testResultsSelector";

const TestResultsList = () => {
  const testResults = useSelector(getTestResultsWithPatientDetails);

  const testResultRows = (results) => {
    if (results.length === 0) {
      return;
    }

    return results.map((result) => (
      <tr key={result.patientId}>
        <th scope="row">{result.displayName}</th>
        <td>{result.patientId}</td>
        <td>{moment(result.dateTested).format("MMM DD YYYY")}</td>
        <td>{result.result}</td>
      </tr>
    ));
  };

  let rows = testResultRows(testResults);
  return (
    <React.Fragment>
      <main className="prime-home">
        <div className="grid-container">
          <div className="grid-row">
            <div className="prime-container usa-card__container">
              <div className="usa-card__header">
                <h2> Test Results </h2>
              </div>
              <div className="usa-card__body">
                <table className="usa-table usa-table--borderless width-full">
                  <thead>
                    <tr>
                      <th scope="col">{PATIENT_TERM_CAP} Name</th>
                      <th scope="col">Unique ID</th>
                      <th scope="col">Date of Test</th>
                      <th scope="col">Result</th>
                    </tr>
                  </thead>
                  <tbody>{rows}</tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </main>
    </React.Fragment>
  );
};

export default TestResultsList;
