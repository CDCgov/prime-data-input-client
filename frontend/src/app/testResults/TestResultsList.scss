@import "../../scss/settings/app-settings.scss";

.sr-test-results-list {
  .sr-more-test-results {
    display: flex;
    justify-content: center;
    margin: 1em 0;
  }

  .sr-test-result-row th,
  .sr-test-result-row td {
    /* override USWDS force-white */
    background: inherit;
  }

  .sr-test-result-row:hover {
    background: $theme-color-base-lightest;
  }

  .sr-test-result-row--removed {
    text-decoration: line-through;
  }

  .usa-form-group {
    margin-top: 0;
  }

  .prime-search-container {
    margin-bottom: 0;

    .prime-search-input {
      margin-bottom: 0;

      .usa-input {
        margin-top: 0.5rem;
        margin-bottom: 0;
        height: 2.5rem;
      }

      .usa-button {
        margin-top: 0.5rem;
        height: 2.5rem;
      }
    }
  }

  .results-dropdown {
    position: relative;
    top: 0;
  }
}

.sr-showing-results-on-page {
  margin-left: 2rem;
  color: #555;
  font-size: 50%;
  vertical-align: middle;
  font-weight: normal;
}
