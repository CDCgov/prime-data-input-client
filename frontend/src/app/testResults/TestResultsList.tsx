import { gql, useQuery } from "@apollo/client";
import React, { useEffect } from "react";
import moment from "moment";
import {
  useAppInsightsContext,
  useTrackEvent,
} from "@microsoft/applicationinsights-react-js";

import { PATIENT_TERM_CAP } from "../../config/constants";
import { displayFullName } from "../utils";

export const testResultQuery = gql`
  query GetFacilityResults($facilityId: String!) {
    testResults(facilityId: $facilityId) {
      internalId
      dateTested
      result
      deviceType {
        internalId
        name
      }
      patient {
        internalId
        firstName
        middleName
        lastName
        lookupId
      }
    }
  }
`;

interface Props {
  activeFacilityId: string;
}

const TestResultsList: any = ({ activeFacilityId }: Props) => {
  const appInsights = useAppInsightsContext();
  const trackFetchTestResults = useTrackEvent(
    appInsights,
    "Fetch Test Results",
    {}
  );

  useEffect(() => {
    trackFetchTestResults({});
  }, [trackFetchTestResults]);
  const { data, loading, error } = useQuery(testResultQuery, {
    variables: { facilityId: activeFacilityId },
    fetchPolicy: "no-cache",
  });

  if (loading) {
    return <p>Loading</p>;
  }
  if (error) {
    appInsights.trackEvent({
      name: "Failed Fetching Tests Results",
    });
    return error;
  }

  const testResultRows = (testResults: any) => {
    if (testResults.length === 0) {
      return;
    }
    const byDateTested = (a: any, b: any) => {
      // ISO string dates sort nicely
      if (a.dateTested === b.dateTested) return 0;
      if (a.dateTested < b.dateTested) return 1;
      return -1;
    };
    // `sort` mutates the array, so make a copy
    return [...testResults].sort(byDateTested).map((r) => (
      <tr key={r.internalId}>
        <th scope="row">
          {displayFullName(
            r.patient.firstName,
            r.patient.middleName,
            r.patient.lastName
          )}
        </th>
        <td>{r.patient.lookupId}</td>
        <td>{moment(r.dateTested).format("lll")}</td>
        <td>{r.result}</td>
        <td>{r.deviceType.name}</td>
      </tr>
    ));
  };

  let rows = testResultRows(data.testResults);
  return (
    <main className="prime-home">
      <div className="grid-container">
        <div className="grid-row">
          <div className="prime-container usa-card__container">
            <div className="usa-card__header">
              <h2> Test Results </h2>
            </div>
            <div className="usa-card__body">
              <table className="usa-table usa-table--borderless width-full">
                <thead>
                  <tr>
                    <th scope="col">{PATIENT_TERM_CAP} Name</th>
                    <th scope="col">Unique ID</th>
                    <th scope="col">Date of Test</th>
                    <th scope="col">Result</th>
                    <th scope="col">Device</th>
                  </tr>
                </thead>
                <tbody>{rows}</tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </main>
  );
};

export default TestResultsList;
