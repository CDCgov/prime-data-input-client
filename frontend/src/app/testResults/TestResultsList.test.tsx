import { render } from "@testing-library/react";
import { DetachedTestResultsList } from "./TestResultsList";

// Data copied from Chrome network window
const testResults = [
  {
    internalId: "3c7998ca-b072-4201-877d-9a1523b0f10c",
    dateTested: "2021-01-04T00:00:00Z",
    result: "UNDETERMINED",
    deviceType: {
      internalId: "f5bd2103-ce04-4290-94a4-ffeea170b0f9",
      name: "LumiraDX",
      __typename: "DeviceType",
    },
    patient: {
      internalId: "633ad102-30f9-4465-b035-bc0c39123fb4",
      firstName: "Cheez",
      middleName: "",
      lastName: "Whizzz",
      lookupId: null,
      __typename: "Patient",
    },
    __typename: "TestResult",
  },
  {
    internalId: "a65bce75-4771-4611-a3b1-9fd6e3d34a1d",
    dateTested: "2021-01-12T17:56:07.403Z",
    result: "NEGATIVE",
    deviceType: {
      internalId: "f5bd2103-ce04-4290-94a4-ffeea170b0f9",
      name: "LumiraDX",
      __typename: "DeviceType",
    },
    patient: {
      internalId: "1eb99bc3-52b1-4bd7-9268-4cfea80a3f8d",
      firstName: "Davis",
      middleName: "",
      lastName: "Melvin",
      lookupId: null,
      __typename: "Patient",
    },
    __typename: "TestResult",
  },
];

describe("TestResultsList", () => {
  it("should render a list of tests", async () => {
    const { container, getByText } = render(
      <DetachedTestResultsList
        activeFacilityId={"fake-facility-1234"}
        data={{ testResults }}
      />
    );
    expect(getByText("Test Results")).toBeInTheDocument();
    expect(getByText("Cheez Whizzz")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
