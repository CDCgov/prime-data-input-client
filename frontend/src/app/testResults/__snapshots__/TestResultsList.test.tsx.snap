// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestResultsList should render a list of tests 1`] = `
<div>
  <main
    class="prime-home"
  >
    <div
      class="grid-container"
    >
      <div
        class="grid-row"
      >
        <div
          class="prime-container usa-card__container sr-test-results-list"
        >
          <div
            class="usa-card__header"
          >
            <h2>
              Test Results
              <span
                class="sr-showing-results-on-page"
              >
                Showing 
                2
                 of
                 
                2
              </span>
            </h2>
          </div>
          <div
            class="usa-card__body"
          >
            <table
              class="usa-table usa-table--borderless width-full"
            >
              <thead>
                <tr>
                  <th
                    scope="col"
                  >
                    Person
                     Name
                  </th>
                  <th
                    scope="col"
                  >
                    Date of Test
                  </th>
                  <th
                    scope="col"
                  >
                    Result
                  </th>
                  <th
                    scope="col"
                  >
                    Device
                  </th>
                  <th
                    scope="col"
                  >
                    Symptoms
                  </th>
                  <th
                    scope="col"
                  >
                    Submitter
                  </th>
                  <th
                    scope="col"
                  >
                    Actions
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr
                  class="sr-test-result-row"
                  data-patient-link="https://simplereport.gov/pxp?plid=68c543e8-7c65-4047-955c-e3f65bb8b58a"
                  title=""
                >
                  <th
                    scope="row"
                  >
                    Cragell, Barb Whitaker
                  </th>
                  <td>
                    Mar 17, 2021 7:27 PM
                  </td>
                  <td>
                    NEGATIVE
                  </td>
                  <td>
                    Abbott IDNow
                  </td>
                  <td>
                    No
                  </td>
                  <td>
                    Arthur A Admin
                  </td>
                  <td>
                    <button
                      aria-expanded="false"
                      aria-haspopup="true"
                      class="rc-menu-button sr-actions-menu"
                    >
                      <svg
                        aria-hidden="true"
                        class="svg-inline--fa fa-ellipsis-h fa-w-16 fa-2x "
                        data-icon="ellipsis-h"
                        data-prefix="fas"
                        focusable="false"
                        role="img"
                        viewBox="0 0 512 512"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M328 256c0 39.8-32.2 72-72 72s-72-32.2-72-72 32.2-72 72-72 72 32.2 72 72zm104-72c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72zm-352 0c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72z"
                          fill="currentColor"
                        />
                      </svg>
                      <span
                        class="usa-sr-only"
                      >
                        More actions
                      </span>
                    </button>
                    <div
                      class="rc-menu-container"
                      role="presentation"
                    />
                  </td>
                </tr>
                <tr
                  class="sr-test-result-row"
                  data-patient-link="https://simplereport.gov/pxp?plid=68c543e8-7c65-4047-955c-e3f65bb8b58a"
                  title=""
                >
                  <th
                    scope="row"
                  >
                    Colleer, Barde X
                  </th>
                  <td>
                    Mar 17, 2021 7:27 PM
                  </td>
                  <td>
                    NEGATIVE
                  </td>
                  <td>
                    Abbott IDNow
                  </td>
                  <td>
                    Yes
                  </td>
                  <td>
                    Ursula User
                  </td>
                  <td>
                    <button
                      aria-expanded="false"
                      aria-haspopup="true"
                      class="rc-menu-button sr-actions-menu"
                    >
                      <svg
                        aria-hidden="true"
                        class="svg-inline--fa fa-ellipsis-h fa-w-16 fa-2x "
                        data-icon="ellipsis-h"
                        data-prefix="fas"
                        focusable="false"
                        role="img"
                        viewBox="0 0 512 512"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M328 256c0 39.8-32.2 72-72 72s-72-32.2-72-72 32.2-72 72-72 72 32.2 72 72zm104-72c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72zm-352 0c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72z"
                          fill="currentColor"
                        />
                      </svg>
                      <span
                        class="usa-sr-only"
                      >
                        More actions
                      </span>
                    </button>
                    <div
                      class="rc-menu-container"
                      role="presentation"
                    />
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <div
            class="usa-card__footer"
          >
            <nav
              aria-label="Pagination"
              class="usa-pagination"
              role="navigation"
            >
              <ol>
                <li>
                  <a
                    aria-label="Page 1"
                    class="is-active"
                    href="/results/1"
                  >
                    <span>
                      1
                    </span>
                  </a>
                </li>
              </ol>
            </nav>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
